spyder/plugins/layout/api.py

```diff
@@ -360,11 +360,9 @@ def set_main_window_layout(self, main_window, dockable_plugins):
             all_plugin_ids.append(plugin.NAME)
 
             # Save currently displayed external plugins
-            if (
-                plugin.NAME in PLUGIN_REGISTRY.external_plugins
-                and plugin.dockwidget.isVisible()
-            ):
-                external_plugins_to_show.append(plugin.NAME)
+            if (plugin.NAME in PLUGIN_REGISTRY.external_plugins):
+                for plugin in dockable_plugins:
+                    external_plugins_to_show.append(plugin.dockwidget.isVisible())
```

Comment:
Hi! I uploaded it this way because here using the "visible = False or True" part works correctly and keeps the plugins when switching Spyder layouts.

Reply:
> Now I will try to use "plugin.NAME" and check if it is "visible = False or True".

Ok, I think that idea could work.

> Sorry for the delay to answer you. I had to do some university stuff.

No problem, I understand.

Reply:
`# Stores boolean values of panels
        external_plugins_value = []

        # External dockable plugins to show after the layout is applied
        external_plugins_to_show = []
        
        # Before applying a new layout all plugins need to be hidden
        for plugin in dockable_plugins:
            all_plugin_ids.append(plugin.NAME)

            # Save currently displayed external plugins
            if (plugin.NAME in PLUGIN_REGISTRY.external_plugins and plugin.dockwidget.isVisible()):
                for plugin in dockable_plugins:
                    external_plugins_value.append(plugin.dockwidget.isVisible())
                
            plugin.toggle_view(False)
            
            # Checks if the value is true
            for value in external_plugins_value:
                if value == True == plugin.dockwidget.isVisible():
                    external_plugins_to_show.append(plugin.NAME)

        plugin.toggle_view(False)`

Reply:
There are some issues with the solution you proposed above:

* This `if` statement

    ```python
    if (plugin.NAME in PLUGIN_REGISTRY.external_plugins and plugin.dockwidget.isVisible()):
        for plugin in dockable_plugins:
            external_plugins_value.append(plugin.dockwidget.isVisible())
    ```

    is adding an entry per dockable plugin to the `external_plugins_value` list with its visibility state. My question is: why is it necessary to iterate over all dockable plugins here?

* This `for` loop

    ```python
    # Checks if the value is true
    for value in external_plugins_value:
        if value == True == plugin.dockwidget.isVisible():
            external_plugins_to_show.append(plugin.NAME)
    ```

    is not adding any plugin to `external_plugins_to_show` because before it you have the command `plugin.toggle_view(False)`, which changes the visibility of `plugin`. In other words `plugin.dockwidget.isVisible()` will always return `False` in the line

    ```python
    if value == True == plugin.dockwidget.isVisible():
    ```

So I don't think it'd work either, sorry.

Reply:
> There are a few problems with the solution you proposed above:
> 
> * This statement`if`
>   ```python
>   if (plugin.NAME in PLUGIN_REGISTRY.external_plugins and plugin.dockwidget.isVisible()):
>       for plugin in dockable_plugins:
>           external_plugins_value.append(plugin.dockwidget.isVisible())
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   is adding a plug-in entry that is in place to the list with its visibility state. My question is: why is it necessary to iterate over all the plugins that fit here?`external_plugins_value`


> There are a few problems with the solution you proposed above:
> 
> * This statement`if`
>   ```python
>   if (plugin.NAME in PLUGIN_REGISTRY.external_plugins and plugin.dockwidget.isVisible()):
>       for plugin in dockable_plugins:
>           external_plugins_value.append(plugin.dockwidget.isVisible())
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   is adding a plug-in entry that is in place to the list with its visibility state. My question is: why is it necessary to iterate over all the plugins that fit here?`external_plugins_value`

Reply:
> * This loop`for`
>   ```python
>   # Checks if the value is true
>   for value in external_plugins_value:
>       if value == True == plugin.dockwidget.isVisible():
>           external_plugins_to_show.append(plugin.NAME)
>   ```
>   
>   
>       
>         
>       
>   
>         
>       
>   
>       
>     
>   is not adding any plugin for because before it you have the command , which changes the visibility of the . In other words, it will always return on the line`external_plugins_to_show``plugin.toggle_view(False)``plugin``plugin.dockwidget.isVisible()``False`
>   ```python
>   if value == True == plugin.dockwidget.isVisible():
>   ```
> 
> So I don't think it would work either, sorry.

Reply:
> So I don't think it would work either, sorry.

Thank you for your help and for your patience.